templatestore/views.py

```diff
@@ -250,6 +232,53 @@ def post_template_view(request):
 
             templates = Template.objects.filter(name=data["name"])
             if not len(templates):
+                if "attributes" not in data:
+                    raise (Exception("Validation: missing field `attributes`"))
+
+                if not len(data["attributes"]):
+                    raise (Exception("Validation: attributes field can not be empty"))
+
+                missing_mandatory_attributes = set(
+                    cfgs[0].attributes.keys()
+                    | ts_settings.TE_TEMPLATE_ATTRIBUTES.keys()
+                ).difference(set(data["attributes"].keys()))
+                if len(missing_mandatory_attributes):
+                    raise (
+                        Exception(
+                            "Validation: missing mandatory attributes `"
+                            + str(missing_mandatory_attributes)
+                            + "` for type `"
+                            + data["type"]
+                            + "`"
+                        )
+                    )
+
+                invalid_valued_attributes = set(
+                    key
+                    for key in cfgs[0].attributes.keys()
+                    if "allowed_values" in cfgs[0].attributes[key]
+                    and data["attributes"][key]
+                    not in cfgs[0].attributes[key]["allowed_values"]
+                )
+
+                invalid_valued_attributes = invalid_valued_attributes | set(
+                    key
+                    for key in ts_settings.TE_TEMPLATE_ATTRIBUTES.keys()
+                    if key in data["attributes"]
+                    and "allowed_values" in ts_settings.TE_TEMPLATE_ATTRIBUTES[key]
+                    and data["attributes"][key]
+                    not in ts_settings.TE_TEMPLATE_ATTRIBUTES[key]["allowed_values"]
+                )
```

Comment:
what is being done here?

Reply:
Here we check that all the attributes (that has a fixed set of values to choose from) have valid values. At first we check that if "allowed_values" are there in the config attributes then is the value given by user from the given set of values or not. Same thing is done for general mandatory attributes (like lob and journey). So invalid_valued_attributes contains a set of attributes that have invalid values. If the set in not empty then we raise an exception that these attributes have invalid values.